### ISSUE TITLE
Console<RT>.readLine repeatedly returning same input without blocking

### ISSUE THREAD

**Issue by timmi-on-rails** [louthy/language-ext#1021]
I modified the original example from https://github.com/louthy/language-ext/wiki/How-to-deal-with-side-effects#ioa which showcases the use of `let`:
```csharp
var computation = from text  in readAllText(inpath)
                  let ntext = Capitalise(text)
                  from _     in writeAllText(outpath, ntext)
                  select unit;
```

to
```csharp
var computation = repeat(from l in Console<RT>.readLine
                  let r = l + l
                  from _ in Console<RT>.writeLine(r)
                  select unit);
```

### Expected
The computation should work as REPL which echos the input twice.

### What happens
When I enter some chars and press Enter, the console is scrolling forever and repeats printing the output without waiting for new user input.

However, the following works:
```csharp
var computation = repeat(from l in Console<RT>.readLine
                  from r in SuccessEff(l + l)
                  from _ in Console<RT>.writeLine(r)
                  select unit);
```

Am I doing something wrong or is this a bug?

**Comment by WarrenCrabb** [comment]
I noticed this same issue when trying the Effects Examples (Queue).  repeat starts an infinite loop.

This issue does not seem to appear in version 4.0.0 but does happen in 4.0.1 and later so almost certainly tied to this commit:  https://github.com/louthy/language-ext/commit/a3925f817ba1f96041b6ec0b2dd3b65a05a3bdbf

Changing this line https://github.com/louthy/language-ext/blob/396fffc8d33d8e3bae00697c4b67f4e4a01a69ab/LanguageExt.Core/Thunks/ThunkSync.Env.cs#L125 back to `var ev = fun(e);`  appears to evaluate the computation correctly and not enter an infinite loop. However, I don't know if that re-introduces the issue raised in the commit.

**Comment by louthy** [comment]
I’ll take a look next week. Thanks for investigating 👍

**Comment by timmi-on-rails** [comment]
Here is another minimal example, which is probably related:

```csharp
Atom<int> counter = Atom(0);

swapEff(counter, i => SuccessEff(i + 1))
    .Select(identity)   // without this line, the code finishes
    .RepeatWhile(i => i < 10)
    .Run();
````

This code runs forever.

**Comment by louthy** [comment]
FYI: I haven't forgotten this, I'm just working on a more robust solution for the `Aff` and `Eff` monads to turn them into DSLs rather than use the Thunks, which are adding a bit too much complexity (and causing edge case issues like this, which in itself was caused by a fix to another issue).

**Comment by louthy** [comment]
This should be resolved in `v4.0.9-beta`.  I need more time on my alternative DSL approach, so I have fallen back to making `Aff` and `Eff` use backing `Func` values instead.  

> NOTE: `Aff` and `Eff` are not memoised like the previous version (which used `Thunk<..>`).  Memoisation causes far too many issues (like the one raised here), and it's arguable that effects should never memoise (due to them interacting with the real world).  So, instead I have added a `Memo()` method for those times when you really want memoisation.

**Comment by timmi-on-rails** [comment]
Thank you very much, I can confirm that both of my examples work as expected with version 4.1.0.
